### ISSUE TITLE
[🐛 Bug]: No way to ignore selenium manager?

### ISSUE THREAD

**Issue by slhck** [SeleniumHQ/selenium#13181]
### What happened?

Consider a simple code (see below) that used to instantiate a ChromeDriver, simply taking the `chromedriver` binary that is available in the `$PATH`.

This used to work with Selenium 4.10.0, but once upgraded to 4.11.0, it no longer "just works", instead, upon first run, installing Selenium Manager, then printing messages like this to the console:

```
Selenium Manager binary found at /Users/werner/Desktop/selenium-test/node_modules/selenium-webdriver/bin/macos/selenium-manager
Driver path: /Users/werner/.nvm/versions/node/v18.18.2/bin/chromedriver
Browser path: /Applications/Google Chrome.app/Contents/MacOS/Google Chrome
```

This breaks several assumptions, e.g., the code not logging when it shouldn't. Also, the first run is significantly slower than needed. It is also not honoring an already-present `chromedriver` in the `$PATH`). In particular, this causes another issue when bundling code, since it now would require bundling the Selenium Manager binary as well (e.g. in WebPack or with rollup).

I have found no way to turn it off though, despite the Selenium Manager documentation claiming that it's an opt-in feature:

> Driver management through Selenium Manager is opt-in for the Selenium bindings. Thus, users can continue managing their drivers manually (putting the driver in the PATH or using system properties)

The documentation of the ServiceBuilder also only says that it will attempt to locate the binary on the PATH: https://www.selenium.dev/selenium/docs/api/javascript/module/selenium-webdriver/chrome_exports_ServiceBuilder.html

Please let users know how to disable that feature and simply use an existing `chromedriver` in the path.

### How can we reproduce the issue?

```shell
const chrome = require("selenium-webdriver/chrome");

const chromeDriverService = new chrome.ServiceBuilder().build();

const opts = new chrome.Options();

const driver = chrome.Driver.createSession(opts, chromeDriverService);

(async () => {
  // load a URL
  await driver.get("https://www.google.com");

  // sleep for a second, then quit
  await driver.sleep(1000);

  await driver.quit();
})();
```

### Relevant log output

```shell
See issue description
```

### Operating System

macOS

### Selenium version

4.11.0 and higher

### What are the browser(s) and version(s) where you see this issue?

Chrome (any version)

### What are the browser driver(s) and version(s) where you see this issue?

ChromeDriver (any version)

**Comment by titusfortner** [comment]
First, please only report bugs based on the latest version of Selenium. We can't go back and fix 4.11, but we can update things in latest. 

So I've done a lot of this work in all of the non JS languages, it's possible that things might not be aligned in JS, we'll investigate. 

Selenium should default to logging at the info level. It's just that most things get logged at the debug level, so you never see anything. In later versions of Selenium we decided that the path information should be debug level, not info level, so you shouldn't see it by default.

Selenium should still use drivers found on PATH, it just uses selenium manager to locate it, instead of having the logic duplicated in all the bindings.

Where did you see that documentation? We need to update it, because you're right that you can no longer put something on PATH if you can't bundle selenium manager and have it work. 

Setting the location of the driver should be possible in JS service class. We don't have an example on our site. @harsha509 do you have example code for this?

**Comment by slhck** [comment]
Thanks for the quick response!

I found the info here: https://www.selenium.dev/documentation/selenium_manager/

The main issue is still present with 4.15.0, so it was just 4.11.0 that broke the behavior.

If I understand you correctly, I cannot forgo the entire Selenium Manager mechanism and instead tell it to use whatever it finds in PATH?

I guess I would somehow have to:

- bundle the binary (impossible to put into one single Node.js file)
- modify its default path (it seems to look in a subfolder)

**Comment by slhck** [comment]
> Setting the location of the driver should be possible in JS service class. We don't have an example on our site. @harsha509 do you have example code for this?

I know that this can be done, see:

```
ServiceBuilder(opt_exe?: string | undefined): chrome.ServiceBuilder
Creates remote.DriverService instances that manage a ChromeDriver server.

@param opt_exe
Path to the server executable to use. If omitted, the builder will attempt to locate the chromedriver on the current PATH.
```

Here is a full repro of the issue: https://github.com/slhck/selenium-webpack-test
I realize this is somewhat of a downstream issue, and can possibly be solved by shipping `selenium_manager` as a file and making Selenium find it, but I hoped it would be possible to simply tell Selenium to use the `chromedriver` and browser that is already present in the `$PATH`.

**Comment by slhck** [comment]
I did some digging through the code. The [`driverLocation` function](https://github.com/SeleniumHQ/selenium/blob/74ae92380900c33c07840a19930ab38e0d459873/javascript/node/selenium-webdriver/common/seleniumManager.js#L72) uses Selenium Manager at all times, even though it "just" returns the driver and browser path. I know there is some complexity in how these are determined, which is the whole purpose of having a common Selenium Manager that hides that.

For the use case where I already have a `chromedriver` in `$PATH`, it seems to suffice to just specify that path when calling the Service Builder:

```javascript
const chromeDriverService = new chrome.ServiceBuilder(chromeDriverBinary).build();
```

When doing that, it seems to ignore Selenium Manager – even in a bundled application. The reason is that the `createSession` function (e.g. in `chromium.js`) then skips the call to Selenium Manager.

So, to summarize, the only solution is to find the `chromedriver` yourself, e.g. with:

```javascript
function getChromeDriverFromPath() {
  const spawnSync = require("child_process").spawnSync;
  const result = spawnSync("which", ["chromedriver"]);
  if (result.status === 0) {
    return result.stdout.toString().trim();
  }
  return undefined;
}

const chromeDriverBinary = getChromeDriverFromPath();
const chromeDriverService = new chrome.ServiceBuilder(chromeDriverBinary).build();
const driver = chrome.Driver.createSession(opts, chromeDriverService); // --> Selenium Manager will be ignored if driver is not undefined
```

I suppose this answers my question but I guess it could be documented in a better way.

**Comment by titusfortner** [comment]
The goal is for users not to need to know how they are getting the right driver/browser, just that they have what they expect and it works. Obviously moving functionality into a separate binary can cause problems for various package managers or bundlers, but hopefully these can be fixed by the projects that do that extra handling.

I think the documentation will make more sense for Selenium 5.0 when it is the default behavior and the most obvious thing is always done without having to consider how it affects backwards compatibility. If there's something we can change in the docs to be more clear in the meantime, please feel free to point it out or create a PR. Thanks.
